File: app/src/main/java/kr/asv/apps/salarytax/fragments/QuickCalculatorFragment.java
package kr.asv.apps.salarytax.fragments;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import kr.asv.apps.salarytax.MoneyTextWatcher;
import kr.asv.apps.salarytax.activities.ReportActivity;
import kr.asv.apps.salarytax.Services;
import kr.asv.calculators.salary.SalaryCalculator;
import kr.asv.shhtaxmanager.R;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link QuickCalculatorFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link QuickCalculatorFragment#newInstance} factory method to
 * create an instance of this fragment.
 *
 */
public class QuickCalculatorFragment extends BaseFragment {
    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment QuickCalculatorFragment.
     */
    public static QuickCalculatorFragment newInstance(String param1, String param2) {
        QuickCalculatorFragment fragment = new QuickCalculatorFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }
    public QuickCalculatorFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_quick_calculator, container, false);
        setFragmentView(view);
        setActionBarTitle(getResources().getString(R.string.activity_title_quick_mode));
        EditText editMoney = (EditText) findViewById(R.id.editMoney_QMode);
        editMoney.addTextChangedListener(new MoneyTextWatcher(editMoney));

        return view;
    }
    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }
    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initEventListener();
    }
    @Override
    public void onDetach() {
        super.onDetach();
    }


    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(Uri uri);
    }

    public void initEventListener()
    {
        // 계산하기 버튼 클릭시
        findViewById(R.id.btnExecute_QMode).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                onClickButtonCalculate(v);// 계산하기 버튼 클릭시
            }
        });

        // 금액 추가 버튼 +천만
        findViewById(R.id.btnPlus1000).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                addInputMoney(10000000);
            }
        });

        // 금액 추가 버튼 +백만
        findViewById(R.id.btnPlus100).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                addInputMoney(1000000);
            }
        });

        // 금액 추가 버튼 +십만
        findViewById(R.id.btnPlus10).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                addInputMoney(100000);
            }
        });

        // 금액 감소 버튼 -천만
        findViewById(R.id.btnMinus1000).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                minusInputMoney(10000000);
            }
        });

        // 금액 감소 버튼 -백만
        findViewById(R.id.btnMinus100).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                minusInputMoney(1000000);
            }
        });

        // 금액 감소 버튼 - 십만
        findViewById(R.id.btnMinus10).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                minusInputMoney(100000);
            }
        });

        // 금액 정정 버튼
        findViewById(R.id.btnClearInput_QM).setOnClickListener(new Button.OnClickListener() {
            @Override
            public void onClick(View v) {
                initInputMoney();
            }
        });
    }
    public void initInputMoney()
    {
        EditText editInputMoney = (EditText) findViewById(R.id.editMoney_QMode);
        editInputMoney.setText("0");
    }
    public void minusInputMoney(int value)
    {
        EditText editInputMoney = (EditText) findViewById(R.id.editMoney_QMode);
        //long inputMoney = getValueEditText(R.id.editMoney_QMode);
        long inputMoney = MoneyTextWatcher.getValue(editInputMoney);
        /*
        if (editInputMoney.getText().length() <= 1) {
            inputMoney = 0;
        } else {
            inputMoney = Integer.parseInt(editInputMoney.getText().toString());
        }
        */
        inputMoney -= value;

        if(inputMoney < 0) inputMoney = 0;
        editInputMoney.setText(String.valueOf(inputMoney));
    }
    public void addInputMoney(int value)
    {
        EditText editInputMoney = (EditText) findViewById(R.id.editMoney_QMode);
        //long inputMoney = getValueEditText(R.id.editMoney_QMode);
        long inputMoney = MoneyTextWatcher.getValue(editInputMoney);
        /*
        if (editInputMoney.getText().length() <= 1) {
            inputMoney = 0;
        } else {
            inputMoney = Integer.parseInt(editInputMoney.getText().toString());
        }
        */
        inputMoney += value;
        editInputMoney.setText(String.valueOf(inputMoney));
    }
    public void onClickButtonCalculate(View v) {
        EditText editInputMoney = (EditText) findViewById(R.id.editMoney_QMode);
        // getMoney
        long inputMoney = 0;
        if (editInputMoney.getText().length() <= 1) {
            return;
        }
        try {
            //inputMoney = getValueEditText(R.id.editMoney_QMode);
            inputMoney = MoneyTextWatcher.getValue(editInputMoney);
        } catch (Exception e) {
            //debug("인트 변환 에러");
            inputMoney = 0;
        }

        /*
         * 연봉기준인지 월급기준인지 구분.
         * 1000만원 이상이면 연봉입력으로 생각하고 계산. (설마 월급이 천만원은 아니겠지)
         */
        boolean annualBasis = false;
        if(inputMoney >= 10000000){
            annualBasis = true;
        }

        //옵션의 기본값
        // 비과세
        long taxExemption = 100000;
        // 부양가족수
        int family = 1;
        // 20세 이하 자녀수
        int child = 0;
        // 퇴직금 포함인지
        boolean includedSeverance = false;


        //환경설정 값 가져오기.
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(getActivity());

        //[퀵계산 설정 사용] 일 때 세부옵션들을 불러온다.
        if(pref.getBoolean("quick_settings_enable",false)){
            family = Integer.parseInt(pref.getString("quick_settings_family","default"));
            child = Integer.parseInt(pref.getString("quick_settings_child","0"));
            taxExemption = Integer.parseInt(pref.getString("quick_settings_tax_exemption","100000"));
            includedSeverance = pref.getBoolean("quick_settings_severance",false);
        }

        SalaryCalculator calculator = Services.getInstance().getCalculator();

        //옵션값 셋팅
        calculator.getOptions().setInputMoney(inputMoney);
        calculator.getOptions().setTaxExemption(taxExemption);
        calculator.getOptions().setFamily(family);
        calculator.getOptions().setChild(child);
        calculator.getOptions().setAnnualBasis(annualBasis);
        calculator.getOptions().setIncludedSeverance(includedSeverance);

        if(pref.getBoolean("rate_settings_enable",false)){
            calculator.getInsurance().getRates().setNationalPension(Double.parseDouble(pref.getString("rate_national_pension","0")));
            calculator.getInsurance().getRates().setHealthCare(Double.parseDouble(pref.getString("rate_health_care","0")));
            calculator.getInsurance().getRates().setLongTermCare(Double.parseDouble(pref.getString("rate_longterm_care","0")));
            calculator.getInsurance().getRates().setEmploymentCare(Double.parseDouble(pref.getString("rate_employment_care","0")));
            //calculator.getInsurance().getRates().setEmploymentCare();
        } else {
            calculator.getInsurance().getRates().init();
        }




        calculator.run();

        //결과 화면 호출
        Intent intent=new Intent(getActivity(),ReportActivity.class);
        startActivity(intent);
    }
}

File: app/src/main/java/kr/asv/apps/salarycalculator/fragments/QuickCalculatorFragment.kt
package kr.asv.apps.salarycalculator.fragments

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.InputFilter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.Toast
import kotlinx.android.synthetic.main.fragment_quick_calculator.*
import kr.asv.androidutils.MoneyTextWatcher
import kr.asv.androidutils.inputfilter.InputFilterLongMinMax
import kr.asv.androidutils.inputfilter.InputFilterMinMax
import kr.asv.apps.salarycalculator.R
import kr.asv.apps.salarycalculator.Services
import kr.asv.apps.salarycalculator.activities.ReportActivity


/**
 * 심플 연봉 계산기
 * 특정 수치를 기준으로, 연봉 or 월급을 나누고 간이 계산한다.
 */
class QuickCalculatorFragment : BaseFragment() {
    private val isDebug = true
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_quick_calculator, container, false)
        setFragmentView(view)
        setActionBarTitle(resources.getString(R.string.nav_menu_quick_calculator))

        val editMoney = findViewById(R.id.idEditTextMoney) as EditText
        editMoney.addTextChangedListener(MoneyTextWatcher(editMoney))
        editMoney.filters = arrayOf<InputFilter>(InputFilterLongMinMax(0, 999999999999))

        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initEventListeners()

        idEditTextMoney.requestFocus()
    }

    override fun onResume() {
        super.onResume()

        // 퀵 계산 모드인 경우에 알려준다.
        if (Services.isCustomRateMode()) {
            Toast.makeText(activity, "'세율설정' 을 사용중입니다. 설정을 취소하시려면 [환경설정 > 고급설정 (세율조정)] 을 변경해주세요.", Toast.LENGTH_LONG).show()
        }
    }

    @Suppress("unused")
    interface OnFragmentInteractionListener {
        fun onFragmentInteraction(uri: Uri)
    }

    /**
     * click event listener 들을 모아둔 메서드.
     */
    private fun initEventListeners() {
        // 계산하기 버튼 클릭시
        btnExecute_QMode.setOnClickListener {
            calculate()// 계산하기 버튼 클릭시
        }

        // 금액 정정 버튼
        btnClearInput_QM.setOnClickListener {
            idEditTextMoney.setText("0")
        }
        
        // 플러스, 마이너스 버튼 이벤트들
        pmButtonsEventListener()

        // 부양가족수 입력 필터
        idFamilyOption.text.filters = arrayOf<InputFilter>(InputFilterMinMax(1, 99))

        // 20세이하 자녀수 입력 필터
        idChildOption.text.filters = arrayOf<InputFilter>(InputFilterMinMax(0, 99))

        // 비과세 입력 필터
        idTaxFreeOption.text.filters = arrayOf<InputFilter>(InputFilterMinMax(0, 999999999))
    }

    /**
     * 증감 관련 버튼 이벤트 리스너
     */
    private fun pmButtonsEventListener(){
        // 금액 추가 버튼 +천만
        idBtnPlus1000.setOnClickListener {
            plusMinusInputMoney(10000000)
        }

        // 금액 추가 버튼 +백만
        idBtnPlus100.setOnClickListener {
            plusMinusInputMoney(1000000)
        }

        // 금액 추가 버튼 +십만
        idBtnPlus10.setOnClickListener {
            plusMinusInputMoney(100000)
        }

        // 금액 감소 버튼 -천만
        idBtnMinus1000.setOnClickListener {
            plusMinusInputMoney(-10000000)
        }

        // 금액 감소 버튼 -백만
        idBtnMinus100.setOnClickListener {
            plusMinusInputMoney(-1000000)
        }

        // 금액 감소 버튼 - 십만
        idBtnMinus10.setOnClickListener {
            plusMinusInputMoney(-100000)
        }
    }

    /**
     * 입력 금액에 증감을 하는 메서드.
     */
    private fun plusMinusInputMoney(value: Int) {
        val editText = findViewById(R.id.idEditTextMoney) as EditText
        //long inputMoney = getValueEditText(R.id.editMoney_QMode);
        var inputMoney = MoneyTextWatcher.getValue(editText)
        inputMoney += value.toLong()

        if (inputMoney < 0) inputMoney = 0
        editText.setText(inputMoney.toString())
    }

    /**
     * 숫자값을 가져올 때, 이상이 있거나 할 때에는 최소값으로 전환시킴.
     * 주의 Int 로 반환함.
     */
    private fun getEditTextNumber(editText: EditText, min: Int): Int{
        return try{
            val i= Integer.parseInt(editText.text.toString())
            if(i < min){
                idFamilyOption.setText(min.toString())
                min
            } else {
                i
            }
        } catch(e: Exception){
            editText.setText(min.toString())
            min
        }
    }

    /**
     * 계산하는 동작
     */
    private fun calculate() {
        // validate checking
        if (idEditTextMoney.text.length <= 1) {
            return
        }

        // 입력 금액
        val inputMoney: Long = MoneyTextWatcher.getValue(idEditTextMoney)
        if(inputMoney <= 0){
            idEditTextMoney.setText("0")
        }

        // 부양 가족수
        // 1 보다 큰 값이어야 함. (물론 입력에서 체크하도록 해야함. 여기는 마지노선 체크)
        val family: Int = getEditTextNumber(idFamilyOption,1)

        // 20세 이하 자녀수
        val child: Int = getEditTextNumber(idChildOption,0)

        // 연봉기준인지, 월급기준인지 구분
        val annualBasis = raMoneyYearly.isChecked

        // 비과세 금액
        val taxFree: Long = MoneyTextWatcher.getValue(idTaxFreeOption)
        if(taxFree <= 0){
            idTaxFreeOption.setText("0")
        }

        //옵션의 기본값
        val includedSeverance = checkSeverance.isChecked // 퇴직금 포함인지

        // 비과세보다 적은 경우는 그냥 계산하지 말기.
        if(inputMoney < taxFree){
            return
        }
        
        // 실수령액 계산기 클래스
        val calculator = Services.calculator

        // 옵션값 셋팅
        val options = calculator.options
        options.inputMoney = inputMoney
        options.taxExemption = taxFree
        options.family = family
        options.child = child
        options.isAnnualBasis = annualBasis
        options.isIncludedSeverance = includedSeverance
        options.isIncomeTaxCalculationDisabled = true

        // 세율 정보 가져오기
        // 커스텀 세율 모드인 경우에는 커스텀 설정을 따름.
        if (Services.isCustomRateMode()) {
            val cKeys = Services.AppPrefKeys.CustomRates
            val rates = calculator.insurance.rates
            rates.nationalPension = Services.getAppPrefValue(cKeys.nationalPension) as Double
            rates.healthCare = Services.getAppPrefValue(cKeys.healthCare) as Double
            rates.longTermCare = Services.getAppPrefValue(cKeys.longTermCare) as Double
            rates.employmentCare = Services.getAppPrefValue(cKeys.employmentCare) as Double
        } else {
            Services.setInsuranceRatesToDefault()
        }

        //
        calculator.prepare()

        // 소득세 계산 (데이터베이스 에서 읽어오기)
        val incomeTaxDao = Services.getIncomeTaxDao()
        incomeTaxDao.isDebug = isDebug
        val earnedIncomeTax = incomeTaxDao.getValue(calculator.salary.basicSalary, family, "201802")
        debug("incomeTax:",earnedIncomeTax)

        // 실수령액 계산
        calculator.run(earnedIncomeTax)

        //결과 화면 호출
        val intent = Intent(activity, ReportActivity::class.java)
        startActivity(intent)
    }

    /**
     * 디버깅 메서드
     */
    @Suppress("unused", "UNUSED_PARAMETER", "SameParameterValue")
    private fun debug(msg: Any, msg2 : Any = "") {
        if (isDebug) {
            Services.debugLog("QuickCalculatorFragment", msg, msg2)
        }
    }

    companion object {

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment QuickCalculatorFragment.
         */
        @Suppress("unused")
        fun newInstance(): QuickCalculatorFragment {
            val fragment = QuickCalculatorFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
